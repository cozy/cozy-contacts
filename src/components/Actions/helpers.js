/**
 * Make object of actions for SelectionBar component
 *
 * @param {Array} actions - Array of actions
 * @param {Object} options - Options for actions
 * @returns {Object} - Object with actions
 */
export const makeActions = (actions = [], options = {}) => {
  return actions.reduce((acc, action) => {
    const actionMenu = action(options)
    const name = actionMenu.name || action.name

    return { ...acc, [name]: actionMenu }
  }, {})
}
