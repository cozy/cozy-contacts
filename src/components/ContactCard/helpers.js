/**
 * @param {object} props
 * @param {{ name: string, setIsAddressDialogOpen: import('react').Dispatch<import('react').SetStateAction<boolean>> }} opts
 * @returns {object}
 */
export const handleContactFieldInputProps = (props, opts) => {
  const { name, setIsAddressDialogOpen } = opts
  if (name.includes('address')) {
    return {
      ...props,
      onClick: () => setIsAddressDialogOpen(true),
      inputProps: { className: 'u-ta-left u-spacellipsis u-h-100' }
    }
  }

  return props
}
