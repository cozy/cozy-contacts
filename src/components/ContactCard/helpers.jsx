import React from 'react'

import Icon from 'cozy-ui/transpiled/react/Icon'
import BottomIcon from 'cozy-ui/transpiled/react/Icons/Bottom'
import InputAdornment from 'cozy-ui/transpiled/react/InputAdornment'

/**
 * @param {object} props
 * @param {{ name: string, setIsAddressDialogOpen: import('react').Dispatch<import('react').SetStateAction<boolean>> }} opts
 * @returns {object}
 */
export const handleContactFieldInputProps = (
  props,
  { name, setIsAddressDialogOpen, setIsRelatedContactDialogOpen }
) => {
  if (name.includes('address')) {
    return {
      ...props,
      onClick: () => setIsAddressDialogOpen(true)
    }
  }

  if (name.includes('relatedContact')) {
    return {
      ...props,
      onClick: () => setIsRelatedContactDialogOpen(true),
      InputProps: {
        endAdornment: (
          <InputAdornment position="end">
            <Icon icon={BottomIcon} color="var(--iconTextColor)" />
          </InputAdornment>
        )
      }
    }
  }

  return props
}
