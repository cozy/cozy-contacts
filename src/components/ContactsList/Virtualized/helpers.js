export const makeColumns = ({ t, isMobile }) => {
  const mobileColumns = [
    {
      id: 'fullname',
      label: t('fields.familyName'),
      width: '100%',
      sortable: false
    },
    {
      id: 'actions',
      width: 40,
      label: '',
      textAlign: 'center',
      disableClick: true,
      sortable: false
    }
  ]

  if (isMobile) {
    return mobileColumns
  }

  return [
    {
      id: 'fullname',
      disablePadding: true,
      label: t('fields.familyName'),
      width: '45%',
      sortable: false
    },
    {
      id: 'email.[0].address',
      label: t('fields.email'),
      textAlign: 'left',
      width: '30%',
      sortable: false
    },
    {
      id: 'phone.[0].number',
      label: t('fields.phone'),
      textAlign: 'left',
      width: 'auto',
      sortable: false
    },
    {
      id: 'actions',
      width: 40,
      label: '',
      textAlign: 'center',
      disableClick: true,
      sortable: false
    }
  ]
}
