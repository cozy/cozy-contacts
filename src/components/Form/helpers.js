export const makeCustomLabel = (value, t) => {
  const { type, label } = JSON.parse(value)

  const firstString = type || ''
  const secondString = label
    ? type
      ? ` (${t(`label.${label}`)})`.toLowerCase()
      : `label.${label}`
    : ''

  return firstString + secondString || null
}

export const makeInitialCustomValue = (name, value) => {
  // gender input doesn't support custom label
  if (!name || !value || name === 'gender') return undefined

  const valueObj = JSON.parse(value)

  // for backwards compatiblity - historically there is only type and no label
  if (valueObj.type && !valueObj.label) {
    return JSON.stringify({ type: valueObj.type })
  }

  // for phone label
  if (name.includes('phoneLabel')) {
    // but unsupported one
    if (!['cell', 'voice', 'fax'].includes(valueObj.type)) {
      return JSON.stringify({ type: valueObj.type, label: valueObj.label })
    }

    // we don't want to create a custom label if supported
    return undefined
  }

  // at this point if label and type are both present, it's a custom label
  if (valueObj.type && valueObj.label) {
    return JSON.stringify({ type: valueObj.type, label: valueObj.label })
  }
}
