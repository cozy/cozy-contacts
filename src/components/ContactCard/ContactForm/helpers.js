import isEqual from 'lodash/isEqual'
import merge from 'lodash/merge'

import { Association } from 'cozy-client'
import { makeDisplayName } from 'cozy-client/dist/models/contact'

import contactToFormValues from './contactToFormValues'
import { DOCTYPE_CONTACTS } from '../../../helpers/doctypes'

/**
 * @param {object} [item] - Contact attribute
 * @returns {string} Stringified object
 */
export const makeItemLabel = item => {
  if (!item) return undefined

  const res =
    item.label || item.type
      ? JSON.stringify({ type: item.type, label: item.label })
      : undefined

  return res
}

/**
 *
 * @param {string} [itemLabel] - Value of the label for a contact attribute
 * @returns {{ type?: string, label?: string }}
 */
export const makeTypeAndLabel = itemLabel => {
  if (!itemLabel) {
    return { type: undefined, label: undefined }
  }

  const itemLabelObj = JSON.parse(itemLabel)

  const res = { type: itemLabelObj.type, label: itemLabelObj.label }

  return res
}

/**
 * @param {object} addressField
 * @returns {boolean} True if addressField has extended address
 */
export const hasExtendedAddress = addressField => {
  if (!addressField) return false
  const extendedAddressKeys = [
    'addresslocality',
    'addressbuilding',
    'addressstairs',
    'addressfloor',
    'addressapartment',
    'addressentrycode'
  ]
  return Object.keys(addressField).some(ext =>
    extendedAddressKeys.includes(ext)
  )
}

export const moveToHead = shouldBeHead => items =>
  items.reduce((arr, v) => (shouldBeHead(v) ? [v, ...arr] : [...arr, v]), [])

export const movePrimaryToHead = moveToHead(v => v?.primary)

export const createAddress = ({ address, oldContact, t }) => {
  return address
    ? address
        .filter(val => val && val.address)
        .map((addressField, index) => {
          const oldContactAddress = oldContact?.address?.[index]
          const oldContactFormValues = contactToFormValues(oldContact, t)
            ?.address?.[index]

          const addressHasBeenModified = !isEqual(
            addressField,
            oldContactFormValues
          )

          if (addressHasBeenModified) {
            // Use "code" instead "postcode", to be vcard 4.0 rfc 6350 compliant
            // eslint-disable-next-line no-unused-vars
            const { postcode, ...oldContactAddressCleaned } =
              oldContactAddress || {}
            return {
              // For keep other properties form connectors
              ...oldContactAddressCleaned,
              formattedAddress: addressField.address,
              number: addressField.addressnumber,
              street: addressField.addressstreet,
              code: addressField.addresscode,
              city: addressField.addresscity,
              region: addressField.addressregion,
              country: addressField.addresscountry,
              ...(hasExtendedAddress(addressField) && {
                extendedAddress: {
                  ...oldContactAddressCleaned.extendedAddress,
                  locality: addressField.addresslocality,
                  building: addressField.addressbuilding,
                  stairs: addressField.addressstairs,
                  floor: addressField.addressfloor,
                  apartment: addressField.addressapartment,
                  entrycode: addressField.addressentrycode
                }
              }),
              ...makeTypeAndLabel(addressField.addressLabel),
              primary: index === 0
            }
          }
          return oldContactAddress
        })
    : []
}

/**
 * @param {(import('../../../types').RelatedContact|undefined)[]} relatedContact - The related contacts array
 * @returns {Record<string, { data: { _id: string, _type: string }[] }>} - The related contacts relationships
 */
export const getRelatedContactRelationships = relatedContact => {
  // Tips filter Boolean to remove undefined value from array when relatedContact is empty (see contactToFormValues)
  const data = relatedContact.filter(Boolean).reduce((acc, curr) => {
    const relationType = curr.relatedContactLabel
      ? JSON.parse(curr.relatedContactLabel).type
      : 'related'

    const existingIndex = acc.findIndex(
      item => item._id === curr.relatedContactId
    )

    if (existingIndex !== -1) {
      acc[existingIndex].metadata.relationTypes = Array.from(
        new Set([...acc[existingIndex].metadata.relationTypes, relationType])
      )
    } else {
      acc.push({
        _id: curr.relatedContactId,
        _type: DOCTYPE_CONTACTS,
        metadata: {
          relationTypes: [relationType]
        }
      })
    }
    return acc
  }, [])

  // `data` can be empty, you still have to return an object to override the behavior of the cozy-client store, otherwise it will keep the old value, and without refreshing the page, the data will not be up to date in the store and therefore on the interface
  return { related: { data } }
}

/**
 * When changing the type of relationship, it must be ensured that no empty relationship remains.
 * The old and new ones are merged into `formValuesToContact`.
 *
 * @param {import('cozy-client/types/types').IOCozyContact} contact - The contact object with all relationships
 * @returns {import('cozy-client/types/types').IOCozyContact} - The contact object without the related contacts relationships
 */
export const removeRelatedContactRelationships = contact => {
  if (!contact?.relationships) return contact
  const updatedContact = merge({}, contact)

  const relationshipsWithoutRelatedContact = Object.entries(
    updatedContact.relationships
  ).reduce((acc, [relName, relValue]) => {
    if ('related' === relName) {
      acc[relName] = relValue
    }
    return acc
  }, {})

  updatedContact.relationships = relationshipsWithoutRelatedContact

  return updatedContact
}

// TODO : Update dehydrate function to HasMany class in cozy-client
/**
 * This function is used to clean the contact object from the associated data
 * cozy-client dehydrates the document before saving it (via the `HasMany` method), but by doing it manually, we ensure that all hydrated relationships in the document (and without data of course) are not saved in the `relationships` of the document, which adds unnecessary data.
 *
 * @param {import('cozy-client/types/types').IOCozyContact} contact - The contact object with associated data
 * @returns {import('cozy-client/types/types').IOCozyContact} - The contact object without associated data
 */
export const removeAsscociatedData = contact => {
  if (!contact) return {}
  return Object.entries(contact).reduce((cleanedContact, [key, value]) => {
    // Add `groups` condition to keep the old implementation functional, see below
    if (!(value instanceof Association) || key === 'groups') {
      cleanedContact[key] = value
    }
    return cleanedContact
  }, {})
}

/**
 * @param {import('cozy-client/types/types').IOCozyContact} contact
 * @returns {import('../../../types').RelatedContact[]}
 */
export const makeRelatedContact = contact => {
  if (
    !(contact.related instanceof Association) ||
    !contact.relationships?.related
  ) {
    return [undefined]
  }

  const relatedData = contact.related.data.reduce((acc, curr) => {
    // Use `makeDisplayName` because if the contact is newly created, it has no `displayName` attribute. (Creation of a contact when selecting a linked contact)
    acc[curr._id] = curr.displayName || makeDisplayName(curr)
    return acc
  }, {})

  const res = contact.relationships.related.data.flatMap(item => {
    return item.metadata.relationTypes.map(type => {
      return {
        relatedContactId: item._id,
        relatedContact: relatedData[item._id],
        relatedContactLabel: makeItemLabel({
          type: type === 'related' ? '' : type
        })
      }
    })
  })

  // Useful because a contact always has at least the `related` relationships (see `getRelatedContactRelationships`)
  return res.length > 0 ? res : [undefined]
}
