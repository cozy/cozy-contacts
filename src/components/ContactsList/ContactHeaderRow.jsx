import React from 'react'
import PropTypes from 'prop-types'
import { scrollToElement } from '../../helpers/scrollToElement'

const ContactHeaderRow = props => (
  <div className="divider">
    <div className="divider__header">{props.header}</div>
    <div className="divider__nav">
      {props.letters.map(letter => (
        <button
          key={letter}
          onClick={() => scrollToElement(`CategorizedList-cat-${letter}`)}
          className={`divider__nav__item ${
            props.header === letter ? '-selected' : ''
          }`}
        >
          {letter}
        </button>
      ))}
    </div>
  </div>
)

ContactHeaderRow.propTypes = {
  header: PropTypes.string.isRequired,
  letters: PropTypes.arrayOf(PropTypes.string).isRequired
}
export default ContactHeaderRow
