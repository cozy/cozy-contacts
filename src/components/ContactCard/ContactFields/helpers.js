import { makeCustomLabel } from '../../Form/helpers'

const makeCustomOrSupportedLabel = ({ type, value, hasPrefix, t }) => {
  if (value.type) {
    return makeCustomLabel(JSON.stringify(value), t)
  }

  if (value.label) {
    return hasPrefix
      ? t(`label.${type}.${value.label}`)
      : t(`label.${value.label}`)
  }

  return null
}

export const makeTLabel = ({ type, value, t, polyglot }) => {
  const hasPrefix = ['phone', 'address'].includes(type)

  if (hasPrefix) {
    if (type === 'phone') {
      const keyToTranslate = `label.${type}.${value.type}-${value.label}`

      return polyglot.has(keyToTranslate)
        ? t(keyToTranslate)
        : makeCustomLabel(JSON.stringify(value), t)
    }

    // has prefix but is not phone
    return makeCustomOrSupportedLabel({ type, value, hasPrefix, t })
  }

  // has no prefix
  return makeCustomOrSupportedLabel({ type, value, hasPrefix, t })
}
