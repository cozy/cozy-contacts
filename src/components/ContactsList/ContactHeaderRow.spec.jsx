import { mount } from 'enzyme'
import React from 'react'
import renderer from 'react-test-renderer'

import ContactHeaderRow from './ContactHeaderRow'

// Mocking scrollToElement
import { scrollToElement } from '../../helpers/scrollToElement'
jest.mock('../../helpers/scrollToElement')

describe('ContactHeaderRow', () => {
  it('should match the ContactHeaderRow snapshot', () => {
    const tree = renderer
      .create(<ContactHeaderRow header="A" letters={['A', 'B', 'C']} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should call scrollToElement on click', () => {
    const wrapper = mount(
      <ContactHeaderRow header="A" letters={['A', 'B', 'C']} />
    )
    // If we click on B
    wrapper
      .find('button')
      .at(1)
      .simulate('click')
    // scrollToElement should be called with DOM id corresponding to B
    expect(scrollToElement.mock.calls).toEqual([['CategorizedList-cat-B']])
  })
})
