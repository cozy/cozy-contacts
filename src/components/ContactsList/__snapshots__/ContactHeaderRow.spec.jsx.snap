// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactHeaderRow should match the ContactHeaderRow snapshot 1`] = `
<div
  className="divider"
>
  <div
    className="divider__header"
  >
    A
  </div>
  <div
    className="divider__nav"
  >
    <button
      className="divider__nav__item -selected"
      onClick={[Function]}
    >
      A
    </button>
    <button
      className="divider__nav__item "
      onClick={[Function]}
    >
      B
    </button>
    <button
      className="divider__nav__item "
      onClick={[Function]}
    >
      C
    </button>
  </div>
</div>
`;
